modules/programs/waybar.nix

```diff
@@ -0,0 +1,388 @@
+    "wlr/taskbar"
+    "idle_inhibitor"
+    "memory"
+    "cpu"
+    "clock"
+    "disk"
+    "tray"
+    "network"
+    "backlight"
+    "pulseaudio"
+    "mpd"
+    "temperature"
+    "bluetooth"
+    "battery"
+  ];
+
+  isValidCustomModuleName = x:
+    elem x defaultModuleNames || (hasPrefix "custom/" x && stringLength x > 7);
+
+  margins = let
+    mkMargin = name: {
+      "margin-${name}" = mkOption {
+        type = nullOr int;
+        default = null;
+        example = 10;
+        description = "Margins value without unit.";
+      };
+    };
+    margins = map mkMargin [ "top" "left" "bottom" "right" ];
+  in foldl mergeAttrs { } margins;
+
+  waybarBarConfig = submodule {
+    options = {
+      layer = mkOption {
+        type = nullOr (enum [ "top" "bottom" ]);
+        default = null;
+        description = ''
+          Decide if the bar is displayed in front (<code>"top"</code>) of the windows or behind (<code>"bottom"</code>).'';
+        example = "top";
+      };
+
+      output = mkOption {
+        type = nullOr (either str (listOf str));
+        default = null;
+        example = literalExample ''
+          [ "DP-1" "!DP-2" "!DP-3" ]
+        '';
+        description = ''
+          Specifies on which screen this bar will be displayed.
+          Exclamation mark(!) can be used to exclude specific output.
+        '';
+      };
+
+      position = mkOption {
+        type = nullOr (enum [ "top" "bottom" "left" "right" ]);
+        default = null;
+        example = "right";
+        description = "Bar position relative to the output.";
+      };
+
+      height = mkOption {
+        type = nullOr ints.unsigned;
+        default = null;
+        example = 5;
+        description =
+          "Height to be used by the bar if possible. Leave blank for a dynamic value.";
+      };
+
+      width = mkOption {
+        type = nullOr ints.unsigned;
+        default = null;
+        example = 5;
+        description =
+          "Width to be used by the bar if possible. Leave blank for a dynamic value.";
+      };
+
+      modules-left = mkOption {
+        type = nullOr (listOf str);
+        default = null;
+        description = "Modules that will be displayed on the left.";
+        example = literalExample ''
+          [ "sway/workspaces" "sway/mode" "wlr/taskbar" ]
+        '';
+      };
+
+      modules-center = mkOption {
+        type = nullOr (listOf str);
+        default = null;
+        description = "Modules that will be displayed in the center.";
+        example = literalExample ''
+          [ "sway/window" ]
+        '';
+      };
+
+      modules-right = mkOption {
+        type = nullOr (listOf str);
+        default = null;
+        description = "Modules that will be displayed on the right.";
+        example = literalExample ''
+          [ "mpd" "custom/mymodule#with-css-id" "temperature" ]
+        '';
+      };
+
+      # modules = modules-options;
+      modules = mkOption {
+        type = attrs;
```

Comment:
Would it be possible to use some `attrsOf` type to make merging work ok?

Reply:
Since modules configurations are within a `listOf`, I don't know how one could merge two different sets.

Does `attrsOf str` allow the user to use any type they wish?

``` nix
{
  modules = {
    "sway/window" = { max-length = 120; };
    "sway/mode" = { tooltip = false; };
  };
}
```

Reply:
Fixed by changing to `attrsOf unspecified`